File: cwsfroster/SettingsViewController.m
//
//  SettingsViewController.m
//  cwsfroster
//
//  Created by Bobby Ren on 8/29/14.
//  Copyright (c) 2014 Bobby Ren. All rights reserved.
//

#import "SettingsViewController.h"
#import "MBProgressHUD.h"
#import "UIImage+Resize.h"
#import "Parse/Parse.h"

#define SECTION_TITLES @[@"About", @"My organization", @"My account", @"Feedback", @"Logout"]
@interface SettingsViewController ()

@end

@implementation SettingsViewController

-(void)didClickClose:(id)sender {
    [self.navigationController dismissViewControllerAnimated:YES completion:nil];
}
#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return SECTION_TITLES.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    
    // Configure the cell...
    cell.textLabel.text = SECTION_TITLES[indexPath.row];
    
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    int row = indexPath.row;
    if (row == 0) {
        // about
        NSString *message = [NSString stringWithFormat:@"Version %@\nCopyright RenderApps, LLC 2017\n", VERSION];
        [UIAlertView alertViewWithTitle:@"About RollCall" message:message];
    }
    else if (row == 1) {
        // my company
        NSString *title = [Organization current].name;
        NSString *message = @"Please select from the following options";
        [UIAlertView alertViewWithTitle:title message:message cancelButtonTitle:@"Cancel" otherButtonTitles:@[@"Change name", @"Change logo"] onDismiss:^(int buttonIndex) {
            if (buttonIndex == 0) {
                [self goToUpdateName];
            }
            else if (buttonIndex == 1) {
                NSLog(@"Change logo");
                [self setupCameraHelper];
                [self goToUpdateLogo];
            }
        } onCancel:nil];
    }
    else if (row == 2) {
        // my account
        NSString *title = nil;
        NSString *message = @"Please select from the following options";
        [UIAlertView alertViewWithTitle:title message:message cancelButtonTitle:@"Cancel" otherButtonTitles:@[@"Change login", @"Change email"] onDismiss:^(int buttonIndex) {
            if (buttonIndex == 0) {
                [self goToUpdateUsername];
            }
            else if (buttonIndex == 1) {
                NSLog(@"Change logo");
                [self goToUpdateUserEmail];
            }
        } onCancel:nil];    }
    else if (row == 3) {
        // feedback
        [self goToFeedback];
    }
    else if (row == 4) {
        // logout
        [AuthService logout];
        [self notifyForLogoutInSuccess];
    }
}

-(void)goToFeedback {
    if ([MFMailComposeViewController canSendMail]){
        NSString *title = @"RollCall feedback";
        NSString *message = [NSString stringWithFormat:@"\n\nOrganization: %@\nVersion %@", [Organization current].name, VERSION];
        MFMailComposeViewController *composer = [[MFMailComposeViewController alloc] init];
        composer.mailComposeDelegate = self;
        [composer setSubject:title];
        [composer setToRecipients:@[@"bobby+rollcall@renderapps.io"]];
        [composer setMessageBody:message isHTML:NO];

        [self presentViewController:composer animated:YES completion:nil];
    }
    else {
        [UIAlertView alertViewWithTitle:@"Currently unable to send email" message:@"Please make sure email is available"];
    }
}

-(void)goToUpdateName {
    NSLog(@"Change name");
    NSString *title = [NSString stringWithFormat:@"Organization: %@.", [[Organization current] name]];
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:title message:@"Please enter new organization name" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"Update", nil];
    alert.alertViewStyle = UIAlertViewStylePlainTextInput;
    alert.tag = 1;
    [alert show];

    [ParseLog logWithTypeString:@"UpdateOrganizationName" title:[[Organization current] objectId] message:nil params:nil error:nil];
}

-(void)goToUpdateUsername {
    NSLog(@"Change name");
    NSString *message = _currentUser.username?[NSString stringWithFormat:@"Your current login is %@", _currentUser.username]:nil;
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Please update your login name" message:message delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"Update", nil];
    alert.alertViewStyle = UIAlertViewStylePlainTextInput;
    alert.tag = 2;
    [alert show];

    [ParseLog logWithTypeString:@"UpdateOrganizationLogin" title:[[Organization current] objectId] message:nil params:nil error:nil];
}

-(void)goToUpdateUserEmail {
    NSLog(@"Change email");
    NSString *message = _currentUser.email?[NSString stringWithFormat:@"Your current email is %@", _currentUser.email]:nil;
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Please enter your email" message:message delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"Update", nil];
    alert.alertViewStyle = UIAlertViewStylePlainTextInput;
    UITextField *text = [alert textFieldAtIndex:0];
    text.keyboardType = UIKeyboardTypeEmailAddress;
    alert.tag = 3;
    [alert show];

    [ParseLog logWithTypeString:@"UpdateOrganizationEmail" title:[[Organization current] objectId] message:nil params:nil error:nil];
}

-(void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    if (buttonIndex == 0) {
        NSLog(@"index 0 - cancel");
    }
    else {
        NSLog(@"else");
        UITextField * text = [alertView textFieldAtIndex:0];
        if (alertView.tag == 1) {
            Organization *organization = [Organization current];
            organization.name = text.text;
            [organization saveInBackgroundWithBlock:^(BOOL succeeded, NSError * _Nullable error) {
                if (succeeded) {
                    NSLog(@"Saved");
                    [UIAlertView alertViewWithTitle:@"Name saved" message:[NSString stringWithFormat:@"Your organization is now called %@", organization.name]];
                    [self notify:@"organization:name:changed"];
                }
                else {
                    [UIAlertView alertViewWithTitle:@"Error updating name" message:nil];
                }
            }];
        }
        else if (alertView.tag == 2) {
            // username
            _currentUser.username = text.text;
            [_currentUser saveInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
                if (succeeded) {
                    NSLog(@"Saved");
                    [UIAlertView alertViewWithTitle:@"Username saved" message:[NSString stringWithFormat:@"Your login is now %@", _currentUser.username]];
                }
                else {
                    [UIAlertView alertViewWithTitle:@"Error updating name" message:nil];
                }
            }];
        }
        else if (alertView.tag == 3) {
            // username
            _currentUser.email = text.text;
            [_currentUser saveInBackgroundWithBlock:^(BOOL succeeded, NSError *error) {
                if (succeeded) {
                    NSLog(@"Saved");
                    [UIAlertView alertViewWithTitle:@"Email saved" message:[NSString stringWithFormat:@"Your email is now %@", _currentUser.email]];
                }
                else {
                    [UIAlertView alertViewWithTitle:@"Error updating email" message:error.userInfo[@"error"]];
                }
            }];
        }
    }
}

#pragma mark MessageController delegate
- (void)mailComposeController:(MFMailComposeViewController*)controller
          didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error {

    // Notifies users about errors associated with the interface
    switch (result)
    {
        case MFMailComposeResultCancelled:
            //feedbackMsg.text = @"Result: Mail sending canceled";
            break;
        case MFMailComposeResultSaved:
            //feedbackMsg.text = @"Result: Mail saved";
            break;
        case MFMailComposeResultSent:
            //feedbackMsg.text = @"Result: Mail sent";
            [UIAlertView alertViewWithTitle:@"Thanks for your feedback" message:nil];
            break;
        case MFMailComposeResultFailed:
            //feedbackMsg.text = @"Result: Mail sending failed";
            [UIAlertView alertViewWithTitle:@"There was an error sending feedback" message:nil];
            break;
        default:
            //feedbackMsg.text = @"Result: Mail not sent";
            break;
    }
    // dismiss the composer
    [self dismissViewControllerAnimated:YES completion:^{
    }];
}

-(void)showProgress: (NSString *)title {
    if (!progress || !progress.taskInProgress) {
        progress = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    }
    progress.taskInProgress = YES;
    progress.mode = MBProgressHUDModeDeterminateHorizontalBar;
    if (title == nil) {
        progress.labelText = @"Uploading photo";
    } else {
        progress.labelText = title;
    }
}

-(void)updateProgress:(float)percent {
    progress.progress = percent;
}

-(void)hideProgress {
    progress.taskInProgress = NO;
    [progress hide:YES];
    progress = nil;
}

@end

File: cwsfroster/SettingsViewController.swift
//
//  SettingsViewController+Utils.swift
//  cwsfroster
//
//  Created by Bobby Ren on 1/28/17.
//  Copyright © 2017 Bobby Ren. All rights reserved.
//

import Foundation
import UIKit
import RACameraHelper
import MessageUI

class SettingsViewController: UITableViewController {
    var cameraHelper: CameraHelper?
    var alert: UIAlertController?

    // TODO: make case iterable
    enum Sections: String {
        case about = "About"
        case organization = "My organization"
        case profile = "My profile"
        case payment = "Payment settings"
        case feedback = "Feedback"
        case logout = "Logout"
    }
    let SECTION_TITLES: [Sections] = [.about, .organization, .profile, .payment, .feedback, .logout]

    func notifyForLogoutInSuccess() {
        self.notify(.LogoutSuccess, object: nil, userInfo: nil)
    }
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)

        title = "Settings"
        if #available(iOS 13.0, *) {
            let closeButtonItem = UIBarButtonItem(barButtonSystemItem: .close, target: self, action: #selector(didClickClose))
            navigationItem.leftBarButtonItem = closeButtonItem
        } else {
            // Fallback on earlier versions
        }
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Cell")
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    @objc private func didClickClose() {
        navigationController?.dismiss(animated: true, completion: nil)
    }
}

extension SettingsViewController {
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return SECTION_TITLES.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = SECTION_TITLES[indexPath.row].rawValue
        return cell
    }
}

extension SettingsViewController {
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        guard indexPath.row < SECTION_TITLES.count else { return }
        switch SECTION_TITLES[indexPath.row] {
        case .about:
            let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
            let message = "Version \(version)\nCopyright RenderApps, LLC 2018";
            simpleAlert("About RollCall", message: message)
            
        case .organization:
            let title = OrganizationService.shared.currentOrganizationName ?? "Your organization"
            let message = "Please select from the following options"
            let alert = UIAlertController(title: title, message: message, preferredStyle: .actionSheet)
            alert.addAction(UIAlertAction(title: "Change name", style: .default, handler: { (action) in
                self.goToUpdateOrganizationName()
            }))
            alert.addAction(UIAlertAction(title: "Change logo", style: .default, handler: { (action) in
                self.setupCameraHelper()
                self.goToUpdateLogo()
            }))
            alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
            if (UIDevice.current.userInterfaceIdiom == UIUserInterfaceIdiom.pad)
            {
                alert.popoverPresentationController?.sourceView = tableView
                alert.popoverPresentationController?.sourceRect = tableView.rectForRow(at: indexPath)
            }
            present(alert, animated: true, completion: nil)
        case .profile:
            let title = OrganizationService.shared.currentOrganizationName ?? "Your organization"
            let message = "Please select from the following options"
            let alert = UIAlertController(title: title, message: message, preferredStyle: .actionSheet)
            alert.addAction(UIAlertAction(title: "Change email", style: .default, handler: { (action) in
                self.goToUpdateEmail()
            }))
            alert.addAction(UIAlertAction(title: "Change password", style: .default, handler: { (action) in
                self.goToUpdatePassword(nil)
            }))
            alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
            if (UIDevice.current.userInterfaceIdiom == UIUserInterfaceIdiom.pad)
            {
                alert.popoverPresentationController?.sourceView = tableView
                alert.popoverPresentationController?.sourceRect = tableView.rectForRow(at: indexPath)
            }
            present(alert, animated: true, completion: nil)
        case .payment:
            goToPayments()
        case .feedback:
            goToFeedback()
            
        case .logout:
            AuthService.logout()
            notifyForLogoutInSuccess()
        }
    }
}

extension SettingsViewController {
    func goToFeedback() {
        guard let nav = UIStoryboard(name: "Feedback", bundle: nil).instantiateInitialViewController() as? UINavigationController, let _ = nav.viewControllers.first as? FeedbackViewController else { return }
        present(nav, animated: true, completion: nil)
    }

    func goToPayments() {
        guard let nav = UIStoryboard(name: "Stripe", bundle: nil).instantiateInitialViewController() as? UINavigationController, let _ = nav.viewControllers.first as? StripeViewController else { return }
        present(nav, animated: true, completion: nil)
    }

    func goToUpdateOrganizationName() {
        let orgId = OrganizationService.shared.currentOrganizationId
        let orgName = OrganizationService.shared.currentOrganizationName ?? ""
        let title = "Organization: \(orgName)"
        let message = "Please enter new organization name"
        inputPrompt(title: title, message: message) { (newName) in
            guard let name = newName else { return }
            OrganizationService.shared.updateName(name)

            self.simpleAlert("Organization updated", message: "Organization name has been changed to \(name)")
            LoggingService.log(event: .updateOrganizationName, info: ["title": orgId ?? "", "name": name])
            self.notify("organization:name:changed", object: nil, userInfo: nil)
        }
    }
    
    func goToUpdateEmail() {
        guard let email = AuthService.currentUser?.email else { return }
        let title = "Your current login is \(email)"
        let message = "Please enter new login email"
        inputPrompt(title: title, message: message, placeholder: email) { (newEmail) in
            guard let newEmail = newEmail else { return }
            AuthService.currentUser?.updateEmail(to: newEmail, completion: { (error) in
                if let error = error as NSError? {
                    self.simpleAlert("Could not update login", defaultMessage: "Your login could not be updated to \(newEmail)", error: error)
                    LoggingService.log(event: .updateOrganizationEmail, info: ["id": OrganizationService.shared.currentOrganizationId ?? ""], error: error)
                } else {
                    self.simpleAlert("Email login updated", message: "Your new login and email is \(newEmail)")
                    LoggingService.log(event: .updateOrganizationEmail, info: ["id": OrganizationService.shared.currentOrganizationId ?? "", "email": newEmail])
                }
            })
        }
    }

    func goToUpdatePassword(_ entered: String?) {
        let title = entered == nil ? "Update password" : "Confirm password"
        let message = entered == nil ? "Please enter new password" : "Please confirm your new password"
        inputPrompt(title: title, message: message, placeholder: nil) { (newPassword) in
            guard let password = newPassword else { return }
            guard let entered = entered else {
                self.goToUpdatePassword(password)
                return
            }
            guard password == entered else {
                self.simpleAlert("Password mismatch", message: "You must confirm the new password.")
                LoggingService.log(event: .updateOrganizationPassword, info: ["id": OrganizationService.shared.currentOrganizationId ?? "", "error": "Password confirmation did not match"])
                return
            }
            AuthService.currentUser?.updatePassword(to: password, completion: { (error) in
                if let error = error as NSError? {
                    self.simpleAlert("Could not update password", defaultMessage: "Your password could not be updated.", error: error)
                    LoggingService.log(event: .updateOrganizationPassword, info: ["id": OrganizationService.shared.currentOrganizationId ?? ""], error: error)
                } else {
                    self.simpleAlert("Password updated", message: nil)
                    LoggingService.log(event: .updateOrganizationPassword, info: ["id": OrganizationService.shared.currentOrganizationId ?? ""])
                }
            })
        }
    }
    
    fileprivate func inputPrompt(title: String, message: String? = nil, placeholder: String? = nil, completion: @escaping ((String?)->Void)) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addTextField { (textField) in
            textField.placeholder = placeholder
        }
        alert.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
            if let textField = alert.textFields?.first {
                completion(textField.text)
            }
        }))
        self.present(alert, animated: true, completion: nil)
    }
}

extension SettingsViewController: MFMessageComposeViewControllerDelegate, UINavigationControllerDelegate {
    func messageComposeViewController(_ controller: MFMessageComposeViewController, didFinishWith result: MessageComposeResult) {
        dismiss(animated: true, completion: nil)
        switch result {
        case .sent:
            simpleAlert("Feedback sent", message: "Thanks for your feedback!")
        case .failed:
            simpleAlert("Could not send feedback", message: "There was an error sending feedback")
        default:
            return
        }
    }
}

// progress
extension SettingsViewController {
    func showProgress(_ title: String?) {
        let alert = UIAlertController(title: title ?? "Progress", message: nil, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Close", style: .cancel) { [weak self] (action) in
            self?.alert = nil
        })
        
        present(alert, animated: true, completion: nil)
        self.alert = alert
    }
    
    func hideProgress(_ completion:(()->Void)? = nil) {
        if alert == nil {
            completion?()
        } else {
            alert?.dismiss(animated: true, completion: completion)
            alert = nil
        }
    }
    
    func updateProgress(percent: Double = 0) {
        alert?.message = "\(percent * 100)%"
    }
}

extension SettingsViewController: CameraHelperDelegate {
    public func didCancelSelection() {
    }
    
    public func didCancelPicker() {
    }
    
    public func didSelectPhoto(selected: UIImage?) {
        // save image to firebase
        dismiss(animated: false) {
            guard let image = selected else { return }
            self.uploadPhoto(image: image)
        }
    }
    
    func setupCameraHelper() {
        if cameraHelper == nil {
            cameraHelper = CameraHelper()
        }
        cameraHelper?.delegate = self
    }
    
    func goToUpdateLogo() {
        print("UpdateLogo")
        let indexPath = IndexPath(row: SECTION_TITLES.firstIndex(of: .organization)!, section: 0)
        let cell = tableView.cellForRow(at: indexPath)
        cameraHelper?.takeOrSelectPhoto(from: self, fromView: cell)
    }
    
    func uploadPhoto(image: UIImage) {
        guard let id = OrganizationService.shared.currentOrganizationId else { return }
        showProgress("Saving new logo")
        print("FirebaseImageService: uploading org photo for \(id)")
        FirebaseImageService.uploadImage(image: image, type: "organization", uid: id, progressHandler: { (progress) in
            self.updateProgress(percent: progress)
        }) { [weak self] (url) in
            if let url = url {
                OrganizationService.shared.updatePhoto(url)
                print("FirebaseImageService: uploading org photo complete with url \(url)")
                LoggingService.log(event: .updateOrganizationLogo, info: ["title": id])
            } else {
                // failure
                self?.simpleAlert("Upload failed", message: "There was an error uploading a new logo.")
            }
            self?.hideProgress()
        }
    }
}
